src/borrow_tracker/tree_borrows/tree/tests.rs

```diff
@@ -0,0 +1,629 @@
+    /// are "distinguishable" if there exists a sequence of instructions
+    /// that causes UB in the target but not in the source.
+    /// This implementation simply explores the reachable space
+    /// by all sequences of `Event2`.
+    /// This function can be instanciated with `RetX` and `RetY`
+    /// among `!` or `()` to resp. forbid/allow `x`/`y` to lose their
+    /// protector.
+    fn distinguishable<RetX, RetY>(&self, other: &Self, xy_rel: RelPosXY) -> bool
+    where
+        RetX: Exhaustive + 'static,
+        RetY: Exhaustive + 'static,
+    {
+        if self == other { return false; }
+        let mut states = vec![(self.clone(), other.clone())];
+        let mut seen = FxHashSet::default();
+        while let Some(state) = states.pop() {
+            if !seen.insert(state.clone()) { continue; };
+            let (source, target) = state;
+            for evt in <Event2<RetX, RetY>>::exhaustive() {
+                // Generate successor states through events (accesses and protector ends)
+                let Ok(new_source) = source.perform_event2(&evt, xy_rel) else { continue; };
+                let Ok(new_target) = target.perform_event2(&evt, xy_rel) else { return true; };
+                if new_source == new_target { continue; }
+                states.push((new_source, new_target));
+            }
+        }
+        false
+    }
+}
+
+#[test]
+#[should_panic]
+// This is why `Reserved -> Frozen` on foreign read for protected references
+// prevents the insertion of spurious reads: the transition can cause UB in the target
+// later down the line.
+fn reserved_frozen_protected_distinguishable() {
+    let xy_rel = RelPosXY::MutuallyForeign;
+    let source = LocStateProt2 {
+        x: LocStateProt {
+            state: LocationState::new(Permission::new_frozen()).with_access(),
+            prot: true,
+        },
+        y: LocStateProt { state: LocationState::new(Permission::new_reserved(false)), prot: true },
+    };
+    let acc = Access2 { ptr: PtrSelector::X, kind: AccessKind::Read };
+    let target = source.perform_access2(&acc, xy_rel).unwrap();
+    assert!(source.y.state.permission.is_reserved(None));
+    assert!(target.y.state.permission.is_frozen());
+    assert!(!source.distinguishable::<(), ()>(&target, xy_rel))
+}
+
+#[test]
+#[should_panic]
+/// This test in a generalization of `reserved_frozen_protected_distinguishable`
+/// that more exhaustively explores the feasability of inserting a spurious read.
+/// For a number of interleavings of different pieces of code involving
+/// two mutually foreign tags, we test if a spurious read through one
+/// changes the outcome (introduces UB in the target when there is none
+/// in the source).
+fn spurious_read() {
+    // The general code pattern being exhaustively tested is
+    //     [thread 1]             || [thread 2]
+    //     retag x (protect)      || retag y (protect/noprotect)
```

Comment:
Why is this not just
```
retag x
<arbitrary code, some of it potentially on another thread>
optional spurious read
```
If `y` gets retagged before, that's part of the initial state. If `y` gets retagged after, that's part of the arbitrary code.

Reply:
In spirit yes, but I can't think of a way to ensure that this is exhaustive and efficient other than the current solution of limiting the state to exactly two observed pointers ?